orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -1179,11 +1194,65 @@ common::Status GenerateSubgraph(Graph& graph, Node* start_node) {
   return graph.Resolve();
 }
 
+Status RemoveNonDifferentiableEdgeInPartition(
+    Graph& graph,
+    const std::vector<std::string>& non_differentiable_edge_names) {
+
+  assert(!non_differentiable_edge_names.empty());
+
+  // Check a given node args vector and see if it contains non-differentiable edges. If it does, remove it.
+  auto check_node_args = [&](std::vector<NodeArg*> node_args, bool is_inputs, Node& node) {
+      auto it = node_args.begin();
+      auto begin = it;
+      auto end = node_args.end();
+      while (it != end) {
+        if (std::any_of(non_differentiable_edge_names.cbegin(), non_differentiable_edge_names.cend(), [&](std::string substr) {
+              return (*it)->Name().rfind(substr, 0) == 0;
```

Comment:
This is doing a substring test rather than equality.  Is that correct?

[ There is a general point here we should consider -- and this applies across many parts of ORT, not just this PR:  We have a lot of string based tests.  That is fragile given risks of creating strings incorrectly, or having accidental clashes.  It is also inefficient given the need for character-by-character tests (as opposed to being able to use equality on uint64_t IDs for nodes or edges). ]

Reply:
We have Python-based distributed tests now. We should be able to make a non-differentiable on a simple graph and run it from Python.